--- latex.sl.orig	2016-02-21 14:57:43.193910200 +0900
+++ latex.sl	2018-04-08 13:33:58.069898956 +0900
@@ -4491,13 +4491,7 @@
     copy_keymap(MODE, SIMPLE_KEYMAP);
 
     % misc
-    definekey("latex->insert_quote", "\"", MODE);
-    definekey("latex->insert_quote", "`",  MODE);
     definekey("latex->insert_dollar", "$",  MODE);
-    definekey("latex->math_ll", "<", MODE);
-    definekey("latex->math_arrow", ">", MODE);
-    definekey("latex->math_arrow", "-", MODE);
-    definekey("latex->math_arrow", "=", MODE);
 
     foreach $1 ([" ", "~"])
       definekey_textormath("latex->insert_without_spaces", "self_insert_cmd",
@@ -4508,7 +4502,6 @@
     definekey("latex->typo_percent", "%", MODE);
     definekey_textormath("latex->insert_without_spaces",
                          "latex->typo_german_decimal_point", ",", MODE);
-    definekey("latex->typo_dots", ".",  MODE);
 
     % math stuff
     definekey_textormath("self_insert_cmd", "latex->cmd_insert(\"colon\")",
@@ -4517,8 +4510,6 @@
                          ")", MODE);
     definekey_textormath("self_insert_cmd", "latex->math_right_parenthesis",
                          "]", MODE);
-    definekey("latex->insert_limits_char()", "^", MODE);
-    definekey("latex->insert_limits_char()", "_", MODE);
     definekey("latex->newline_with_completion", Key_Shift_Return, MODE);
 
     definekey("latex->unfold_or_newline", Key_Return, MODE);
@@ -4986,9 +4977,11 @@
     % comments:
     dfa_define_highlight_rule("%(.*[^ \t])?", "comment", name);
 
+    % document structures
     dfa_define_highlight_rule("\\documentclass.*}"R, "Qpreprocess", name);
-    dfa_define_highlight_rule("\\begin{.*}({.*})*"R, "preprocess", name);
+    dfa_define_highlight_rule("\\begin{.*}"R, "preprocess", name);
     dfa_define_highlight_rule("\\end{.*}"R, "Qpreprocess", name);
+    dfa_define_highlight_rule("\\section", "preprocess", name);
 
     % % known keywords in curly braces
     % dfa_define_highlight_rule("{article}", "Qstring", name);
@@ -5051,12 +5044,12 @@
     % dfa_define_highlight_rule("^.*\\\\[\\)\\]]", "Qstring", name);
     % dfa_define_highlight_rule("\\\\[\\(\\[].*", "string", name);
 
-    % dfa_define_highlight_rule("\\$.*\\$", "Qnumber", name);
+    dfa_define_highlight_rule("\\$.*\\$", "Qnumber", name);
     % dfa_define_highlight_rule("\\$.*[^ ]", "number", name);
     % %   dfa_define_highlight_rule("^[^\\$]*\\$", "number", name);
 
     % % Fundamental delimiters in the TeX language: {}[]
-    % dfa_define_highlight_rule("[{}\\[\\]]", "delimiter", name);
+    dfa_define_highlight_rule("[{}]", "delimiter", name);
 
     % % \leftX \rightY constructions where X and Y are
     % % one of \| \{ \} [ ]( ) / | .
@@ -5110,11 +5103,21 @@
     dfa_define_highlight_rule("\\[\-,:;!%\$#&_ \|\\/{}~\^'`\.=>]"R,
     			      "keyword1", name);
 
+    % type 3 keywords: definition
+    dfa_define_highlight_rule("\\protected"R, "Qkeyword3", name);
+    dfa_define_highlight_rule("\\def"R, "Qkeyword3", name);
+    dfa_define_highlight_rule("\\gdef"R, "Qkeyword3", name);
+    dfa_define_highlight_rule("\\xdef"R, "Qkeyword3", name);
+    dfa_define_highlight_rule("\\edef"R, "Qkeyword3", name);
+    dfa_define_highlight_rule("\\newcommand"R, "Qkeyword3", name);
+    dfa_define_highlight_rule("\\renewcommand"R, "Qkeyword3", name);
+    dfa_define_highlight_rule("\\DeclareRobustCommand"R, "Qkeyword3", name);
+
     % type 0 keywords: a backslash followed by alpha characters
     dfa_define_highlight_rule("\\["R +
                               str_quote_string(TeX_Command_Chars, "*", '\\')
-                              + "]+", "keyword", name);
-
+                              + "]+", "keyword", name);    
+   
     % % a backslash followed by a single char not covered by one of the
     % % previous rules is probably an error
     % dfa_define_highlight_rule("\\.", "error", name);
@@ -5123,11 +5126,11 @@
     dfa_define_highlight_rule("[~\\^_]", "operator", name);
 
     % numbers
-    dfa_define_highlight_rule("[0-9]([\.,0-9]*[0-9])?"R, "number", name);
+    %dfa_define_highlight_rule("[0-9]([\.,0-9]*[0-9])?"R, "number", name);
     dfa_define_highlight_rule("\$"R, "number", name);
 
     % % macro parameters(#1 #2 etc)
-    % dfa_define_highlight_rule("#[1-9]", "operator", name);
+    dfa_define_highlight_rule("#[1-9]", "number", name);
 
     % quoted strings
     dfa_define_highlight_rule("\"`.*\"'", "Qstring", name); % german
@@ -5142,13 +5145,40 @@
     dfa_define_highlight_rule(".?.?.?\"'", "string", name);
     dfa_define_highlight_rule(".?.?.?''", "string", name);
 
-    dfa_define_highlight_rule("[ \t]+$", "trailing_whitespace", name);
-
     % Workaround to make UTF-8 characters are display correctly, not as
     % <C3><nn> or <E2><nn><nn>
     dfa_define_highlight_rule("\xC2.", "normal", name);
     dfa_define_highlight_rule("\xC3.", "normal", name);
+    dfa_define_highlight_rule("\xC4.", "normal", name);
+    dfa_define_highlight_rule("\xC5.", "normal", name);
+    dfa_define_highlight_rule("\xC6.", "normal", name);
+    dfa_define_highlight_rule("\xC7.", "normal", name);
+    dfa_define_highlight_rule("\xC8.", "normal", name);
+    dfa_define_highlight_rule("\xC9.", "normal", name);
+    dfa_define_highlight_rule("\xCA.", "normal", name);
+    dfa_define_highlight_rule("\xCB.", "normal", name);
+    dfa_define_highlight_rule("\xCD.", "normal", name);
+    dfa_define_highlight_rule("\xCE.", "normal", name);
+    dfa_define_highlight_rule("\xCF.", "normal", name);
+    dfa_define_highlight_rule("\xD0.", "normal", name);
+    dfa_define_highlight_rule("\xD1.", "normal", name);
+    dfa_define_highlight_rule("\xD2.", "normal", name);
+    dfa_define_highlight_rule("\xD3.", "normal", name);
+    dfa_define_highlight_rule("\xD4.", "normal", name);
+    dfa_define_highlight_rule("\xD5.", "normal", name);
+    dfa_define_highlight_rule("\xD6.", "normal", name);
+    dfa_define_highlight_rule("\xD7.", "normal", name);
+    dfa_define_highlight_rule("\xD8.", "normal", name);
+    dfa_define_highlight_rule("\xD9.", "normal", name);
+    dfa_define_highlight_rule("\xDA.", "normal", name);
+    dfa_define_highlight_rule("\xDB.", "normal", name);
+    dfa_define_highlight_rule("\xDD.", "normal", name);
+    dfa_define_highlight_rule("\xDE.", "normal", name);
+    dfa_define_highlight_rule("\xDF.", "normal", name);
+    dfa_define_highlight_rule("\xE0..", "normal", name);
+    dfa_define_highlight_rule("\xE1..", "normal", name);
     dfa_define_highlight_rule("\xE2..", "normal", name);
+    dfa_define_highlight_rule("\xE3\x80\x80", "Qkeyword2", name); % ideographic space
     dfa_define_highlight_rule("\xE3..", "normal", name);
     dfa_define_highlight_rule("\xE4..", "normal", name);
     dfa_define_highlight_rule("\xE5..", "normal", name);
@@ -5156,9 +5186,27 @@
     dfa_define_highlight_rule("\xE7..", "normal", name);
     dfa_define_highlight_rule("\xE8..", "normal", name);
     dfa_define_highlight_rule("\xE9..", "normal", name);
+    dfa_define_highlight_rule("\xEA..", "normal", name);
+    dfa_define_highlight_rule("\xEB..", "normal", name);
+    dfa_define_highlight_rule("\xEC..", "normal", name);
+    dfa_define_highlight_rule("\xED..", "normal", name);
+    dfa_define_highlight_rule("\xEE..", "normal", name);
+    dfa_define_highlight_rule("\xEF\xBC\x88", "Qkeyword4", name); % fullwidth parenthesis
+    dfa_define_highlight_rule("\xEF\xBC\x89", "Qkeyword4", name); % fullwidth parenthesis
+    dfa_define_highlight_rule("\xEF..", "normal", name);
+    dfa_define_highlight_rule("\xF0...", "normal", name);
+    dfa_define_highlight_rule("\xF1...", "normal", name);
+    dfa_define_highlight_rule("\xF2...", "normal", name);
+    dfa_define_highlight_rule("\xF3...", "normal", name);
+    dfa_define_highlight_rule("\xF4...", "normal", name);
+    dfa_define_highlight_rule("\xF5...", "normal", name);
+    dfa_define_highlight_rule("\xF6...", "normal", name);
+    dfa_define_highlight_rule("\xF7...", "normal", name);
 
+    % whitespaces
+    dfa_define_highlight_rule("[ \t]+$", "trailing_whitespace", name);
+   
     % all the rest
-    % Fixme: Why we need this rule?
     dfa_define_highlight_rule(".", "normal", name);
 
     dfa_build_highlight_table(name);
@@ -5260,7 +5308,7 @@
     append_to_hook("_jed_save_buffer_before_hooks", &save_buf_before_hook);
     set_buffer_hook("par_sep", &paragraph_separator);
 
-    WRAP_INDENTS = 1;
+    WRAP_INDENTS = 0;
     define_blocal_var("info_page", "latex");
 
     if ( abbrev_table_p("TeX") )
